[Frontend] Add sagemaker_standards dynamic lora adapter and stateful session management decorators to vLLM OpenAI API server

zhaozuy opened the pull request:
<!-- markdownlint-disable -->

## Purpose
Add AWS SageMaker standards integration to vLLM's OpenAI API server to enable seamless deployment on Amazon SageMaker.

This PR introduces:
  - Integration with the [`model-hosting-container-standards` library](https://github.com/aws/model-hosting-container-standards) to support SageMaker-specific request patterns
  - SageMaker-compatible endpoints for LoRA adapter management (`POST /adapters`, `DELETE /adapters/{adapter_name}`)
  - Stateful session support via the `/invocations` endpoint with SageMaker session headers
  - Automatic adapter injection through `X-Amzn-SageMaker-Adapter-Identifier` header
  - A `--disable-sagemaker-standards` CLI flag for optional disabling

**Changes:**
  - Added `model-hosting-container-standards` dependency to `requirements/common.txt`
  - Applied SageMaker decorators (`@stateful_session_manager()`, `@inject_adapter_id()`) to `/invocations` endpoint in `vllm/entrypoints/openai/api_server.py`
  - Registered SageMaker handlers for `load_lora_adapter` and `unload_lora_adapter` endpoints
  - Added SageMaker bootstrap logic to `build_app()` (can be disabled with `--disable-sagemaker-standards`)
  - Created comprehensive test suite in `tests/sagemaker/`

## Test Plan

Run the new SageMaker test suites:

```bash
# Test LoRA adapter loading/unloading via SageMaker endpoints
pytest tests/sagemaker/test_sagemaker_lora_adapters.py -v

# Test stateful session management
pytest tests/sagemaker/test_sagemaker_stateful_sessions.py -v

# Test sagemaker custom middleware
pytest tests/sagemaker/test_sagemaker_middleware_integration.py -v

# Test sagemaker endpoint overrides
pytest tests/sagemaker/test_sagemaker_handler_overrides.py -v
```

Then run related OpenAI entrypoint tests:

```bash
# Test LoRA adapter loading/unloading via original OpenAI API server endpoints
pytest tests/entrypoints/openai/test_lora_adapters.py -v

# Test OpenAI API server's regular invocations endpoint
pytest -v \
  tests/entrypoints/openai/test_chat.py \
  tests/entrypoints/pooling/openai/test_classification.py \
  tests/entrypoints/pooling/openai/test_embedding.py \
  tests/entrypoints/pooling/openai/test_pooling.py \
  tests/entrypoints/pooling/openai/test_rerank.py \
  tests/entrypoints/pooling/openai/test_score.py \
  -k "test_invocations"
```

## Test Result

New SageMaker-related tests:
```bash
===================================================================== test session starts ======================================================================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /vllm-workspace
plugins: asyncio-1.2.0, typeguard-4.4.4, hypothesis-6.142.4, subtests-0.14.2, schemathesis-3.39.15, anyio-4.11.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 12 items                                                                                                                                             

tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_load_adapter_happy_path[True] PASSED                                                    [  8%]
tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_unload_adapter_happy_path[True] PASSED                                                  [ 16%]
tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_load_adapter_not_found[True] PASSED                                                     [ 25%]
tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_load_adapter_invalid_files[True] PASSED                                                 [ 33%]
tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_unload_nonexistent_adapter[True] PASSED                                                 [ 41%]
tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_invocations_with_adapter[True] PASSED                                                   [ 50%]
tests/sagemaker/test_sagemaker_lora_adapters.py::test_sagemaker_multiple_adapters_load_unload[True] PASSED                                              [ 58%]
tests/sagemaker/test_sagemaker_stateful_sessions.py::test_create_session_badrequest PASSED                                                              [ 66%]
tests/sagemaker/test_sagemaker_stateful_sessions.py::test_close_session_badrequest[nonexistent_session_id-session_id_change0-request_body_change0-session not found] PASSED [ 75%]
tests/sagemaker/test_sagemaker_stateful_sessions.py::test_close_session_badrequest[malformed_close_request-session_id_change1-request_body_change1-None] PASSED [ 83%]
tests/sagemaker/test_sagemaker_stateful_sessions.py::test_close_session_invalidrequest PASSED                                                           [ 91%]
tests/sagemaker/test_sagemaker_stateful_sessions.py::test_session PASSED                                                                                [100%]

====================================================================== warnings summary =======================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305
  /usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305: DeprecationWarning: 
jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference 
resolution errors, directly catch referencing.exceptions.Unresolvable.
    ref_error: type[Exception] = jsonschema.RefResolutionError,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================== 12 passed, 3 warnings in 65.97s (0:01:05) ==========================================================
r
```

Existing tests
```bash
# Tested LoRA adapter loading/unloading via original OpenAI API server endpoints
===================================================================== test session starts ======================================================================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /vllm-workspace
plugins: asyncio-1.2.0, typeguard-4.4.4, hypothesis-6.142.4, subtests-0.14.2, schemathesis-3.39.15, anyio-4.11.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 10 items                                                                                                                                             

tests/entrypoints/openai/test_lora_adapters.py::test_static_lora_lineage[True] PASSED                                                                    [ 10%]
tests/entrypoints/openai/test_lora_adapters.py::test_dynamic_lora_lineage[True] PASSED                                                                   [ 20%]
tests/entrypoints/openai/test_lora_adapters.py::test_dynamic_lora_not_found[True] PASSED                                                                 [ 30%]
tests/entrypoints/openai/test_lora_adapters.py::test_dynamic_lora_invalid_files[True] PASSED                                                             [ 40%]
tests/entrypoints/openai/test_lora_adapters.py::test_dynamic_lora_badrequests[True-test_rank-config_change0-is greater than max_lora_rank] PASSED        [ 50%]
tests/entrypoints/openai/test_lora_adapters.py::test_dynamic_lora_badrequests[True-test_dora-config_change1-does not yet support DoRA] PASSED            [ 60%]
tests/entrypoints/openai/test_lora_adapters.py::test_dynamic_lora_badrequests[True-test_modules_to_save-config_change2-only supports modules_to_save being None
] PASSED [ 70%]
tests/entrypoints/openai/test_lora_adapters.py::test_multiple_lora_adapters[True] PASSED                                                                 [ 80%]
tests/entrypoints/openai/test_lora_adapters.py::test_loading_invalid_adapters_does_not_break_others[True] PASSED                                         [ 90%]
tests/entrypoints/openai/test_lora_adapters.py::test_beam_search_with_lora_adapters[True] PASSED                                                         [100%]

======================================================================= warnings summary =======================================================================
<frozen importlib._bootstrap>:488                                                                                                                              
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305
  /usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305: DeprecationWarning: 
jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference 
resolution errors, directly catch referencing.exceptions.Unresolvable.                           
    ref_error: type[Exception] = jsonschema.RefResolutionError,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================== 10 passed, 3 warnings in 49.66s ================================================================

# Tested OpenAI API server's regular invocations endpoint
===================================================================== test session starts =====================================================================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: /vllm-workspace
plugins: asyncio-1.2.0, typeguard-4.4.4, hypothesis-6.142.4, subtests-0.14.2, schemathesis-3.39.15, anyio-4.11.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 86 items / 77 deselected / 9 selected                                                                                                                

tests/entrypoints/openai/test_chat.py::test_invocations PASSED                                                                                           [ 11%]
tests/entrypoints/pooling/openai/test_classification.py::test_invocations PASSED                                                                         [ 22%]
tests/entrypoints/pooling/openai/test_embedding.py::test_invocations PASSED                                                                              [ 33%]
tests/entrypoints/pooling/openai/test_embedding.py::test_invocations_conversation PASSED                                                                 [ 44%]
tests/entrypoints/pooling/openai/test_pooling.py::test_invocations PASSED                                                                                [ 55%]
tests/entrypoints/pooling/openai/test_pooling.py::test_invocations_conversation PASSED                                                                   [ 66%]
tests/entrypoints/pooling/openai/test_rerank.py::test_invocations PASSED                                                                                 [ 77%]
tests/entrypoints/pooling/openai/test_score.py::TestModel::test_invocations[model0] PASSED                                                               [ 88%]
tests/entrypoints/pooling/openai/test_score.py::TestModel::test_invocations[model1] PASSED                                                               [100%]

======================================================================= warnings summary ======================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305
  /usr/local/lib/python3.12/dist-packages/schemathesis/generation/coverage.py:305: DeprecationWarning: 
jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0. If you wish to catch potential reference 
resolution errors, directly catch referencing.exceptions.Unresolvable.
    ref_error: type[Exception] = jsonschema.RefResolutionError,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================== 9 passed, 77 deselected, 3 warnings in 200.78s (0:03:20) ==================================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

zhaozuy committed:
[Frontend] Add sagemaker standards to vLLM dependencies and add sagemaker lora & stateful session decorators to openai/api_server.py, add tests.

Signed-off-by: Zuyi Zhao <zhaozuy@amazon.com>

simon-mo commented:
I find this to be a bit intrusive, please consider how Anthropic endpoint is implemented https://github.com/vllm-project/vllm/tree/main/vllm/entrypoints/anthropic

zhaozuy committed:
[Frontend] Move LoRA endpoints out of openai api_server and SageMaker specific endpoints out of openai api_server.

Signed-off-by: Zuyi Zhao <zhaozuy@amazon.com>

HappyAmazonian committed:
Add handler override and middleware integration tests

Signed-off-by: Shen Teng <sheteng@amazon.com>

Signed-off-by: Zuyi Zhao <zhaozuy@amazon.com>

HappyAmazonian committed:
Fix line length issues in SageMaker test files

- Split long lines to comply with 88-character limit
- Added noqa comment for unavoidable long import line
- Improved code readability by proper line breaks

Signed-off-by: Shen Teng <sheteng@amazon.com>

simon-mo commented on tests/sagemaker/test_sagemaker_lora_adapters.py:
```diff
@@ -0,0 +1,228 @@
+from ..utils import RemoteOpenAIServer
+
+# any model with a chat template should work here
+MODEL_NAME = "HuggingFaceH4/zephyr-7b-beta"
```
The convention of testing end to end has changed. This is using a legacy model AFIAK

zhaozuy replied:
@simon-mo Is there a recommended model/LoRA adapter combination that we should use instead? I took this model from the existing `/tests/entrypoints/openai/test_lora_adapters.py` file.

simon-mo commented on tests/sagemaker/__init__.py:
this should go into one of the entrypoint tests folder

simon-mo commented on tests/sagemaker/test_sagemaker_handler_overrides.py:
1. It seems the model server isn't cached so it's started each time
2. why are we testing this instead of in model hosting standards?

HappyAmazonian replied:
Hi Simon,
1. Since the customer script gets loaded during the the fastapi app building step, we need to create a new Server for each test case here.
2. Since the model hosting standards doesn't take vllm as a dependency, it has the similar tests with a mocked vllm server. I added these tests here to make sure the user experience work with the actual vllm server.

simon-mo commented on vllm/entrypoints/openai/cli_args.py:
```diff
@@ -169,6 +169,8 @@ class FrontendArgs:
     log. The default of None means unlimited."""
     disable_fastapi_docs: bool = False
     """Disable FastAPI's OpenAPI schema, Swagger UI, and ReDoc endpoint."""
+    disable_sagemaker_standards: bool = False
```
given you added this as default dependencies, there's no need for an extra flag to turn it off?

simon-mo reviewed (changes requested):
My main feedback is the tests is too verbose and written in not efficient way. Please refer to other entrypoint tests for concise and fast testing.

zhaozuy committed:
Fix: improve sagemaker_standard tests by using smaller model for all tests, move folder under entrypoints, remove disable_sagemaker_standards cli flag.

Signed-off-by: Zuyi Zhao <zhaozuy@amazon.com>

zhaozuy committed:
Merge branch 'main' into upstream_sagemaker_standards

zhaozuy committed:
Merge branch 'main' into upstream_sagemaker_standards

zhaozuy committed:
Merge branch 'main' into upstream_sagemaker_standards

robertgshaw2-redhat committed:
Merge branch 'main' into upstream_sagemaker_standards

robertgshaw2-redhat committed:
Merge branch 'main' into upstream_sagemaker_standards

zhaozuy committed:
Pin model-hosting-container-standards version.

Signed-off-by: Zuyi Zhao <zhaozuy@amazon.com>

zhaozuy committed:
Merge branch 'main' into upstream_sagemaker_standards
